lib/kafka_gen_stage/consumer_logic.ex

```diff
@@ -28,71 +27,76 @@ defmodule KafkaGenStage.ConsumerLogic do
   @spec prepare_dispatch(
           buffered_msgs :: :queue.queue(),
           buffered_demand :: non_neg_integer(),
-          transformer :: (msg_tuple(), term() -> {term(), term()}),
-          transformer_state :: term,
-          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()])
+          bulk_transformer :: ([msg_tuple()] -> [msg_tuple()]),
+          high_wm :: non_neg_integer()
         ) ::
           dispatch()
-  def prepare_dispatch(queue, 0, _transformer, state, _bulk_transformer) do
-    {[], :no_ack, 0, queue, state}
+  def prepare_dispatch(queue, 0, bulk_transformer, high_wm) do
+    {transform_bulk([], bulk_transformer, high_wm), :no_ack, 0, queue}
   end
 
-  def prepare_dispatch(queue, demand, transformer, state, bulk_transformer) do
+  def prepare_dispatch(queue, demand, bulk_transformer, high_wm) do
     case :queue.out(queue) do
-      {{:value, msg}, queue} ->
-        {to_send, new_state, to_ack} = transform(msg, transformer, state)
-        new_demand = lower_demand(demand, to_send)
+      {{:value, {offset, _, _, _} = msg}, queue} ->
+        {to_send, to_ack, _demand, queue} =
+          prepare_dispatch(
+            queue,
+            demand - 1,
+            [msg],
+            offset,
+            high_wm
+          )
+
+        new_to_send =
+          to_send
+          |> Enum.reverse()
+          |> transform_bulk(bulk_transformer, high_wm)
 
-        {to_send, to_ack, _demand, queue, new_transformer_state} =
-          prepare_dispatch(queue, new_demand, Enum.reverse(to_send), to_ack, transformer, new_state)
-
-        new_to_send = transform_bulk(to_send, bulk_transformer)
         final_demand = lower_demand(demand, new_to_send)
-        {new_to_send, to_ack, final_demand, queue, new_transformer_state}
+        {new_to_send, to_ack, final_demand, queue}
 
       {:empty, queue} ->
-        {[], :no_ack, demand, queue, state}
+        # proc se ne-ackne, kdyz je queue prazdna? proc se na-ackne predchozi zprava?
```

Comment:
Is this something you'd want to leave in the code? If so, please, translate to english.

Reply:
Not 100% sure, but it seems that message is acked when it's moved from `queue` to `to_send`. Here there's nothing in queue, therefore, everything that was in queue before was already acked (it was already moved from `queue` to `to_send`). So the present logic seems correct to me. Do you have any indication it's not correct?

Reply:
It's how you say it is. Still I added transform_bulk to the following line which was not there before. Please check that too.

Reply:
That seems OK.